### PR TITLE
Create v1alpha2 with bucket quotas

### PR DESCRIPTION
**Pull request by BlaineEXE**
Copy v1alpha1 -> v1alpha2 with no changes. Then...

Add optional quotas to BucketClass:
Focus initially on quotas applied to all buckets created from a
BucketClass. This means that a storage admin may specify limits that
apply to every user of the BucketClass. Users do not have self-service
control over bucket quotas beyond their ability to select amongst some
number of BucketClasses.

### CODE REVIEW

**Comment by BlaineEXE** [comment]
Because this creates v1alpha2, it looks like a lot of changes as a whole. But the 2nd commit shows the actual extent, which is currently quite small. 

https://github.com/kubernetes-sigs/container-object-storage-interface-api/pull/66/commits/a2566de9d593c3a2fa71d97fa9b203878a1b86e5

I don't foresee this being finished yet, but I want to start having initial conversations before addressing things like Bucket status items showing what quotas are in place.

**Comment by xing-yang** [comment]
/hold

This needs more design discussions before merging.

File: apis/objectstorage/v1alpha2/types.go [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+	// bucket. There are 2 possible values:
+	//  - Retain: Indicates that the bucket should not be deleted from the OSP
+	//  - Delete: Indicates that the bucket should be deleted from the OSP
+	//        once all the workloads accessing this bucket are done
+	// +kubebuilder:default:=Retain
+	DeletionPolicy DeletionPolicy `json:"deletionPolicy"`
+
+	// Quotas defines optional quotas applied to all buckets created from this Bucket Class.
+	// +optional
+	Quotas BucketQuotas `json:"quotas,omitempty"`
```
Comment (thotz) [inline apis/objectstorage/v1alpha2/types.go:209]: IMO the BucketClass should have quota based on no of buckets as well

File: apis/objectstorage/v1alpha2/types.go [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+	// AuthenticationType denotes the style of authentication
+	// It can be one of
+	// Key - access, secret tokens based authentication
+	// IAM - implicit authentication of pods to the OSP based on service account mappings
+	AuthenticationType AuthenticationType `json:"authenticationType"`
+
+	// Parameters is an opaque map for passing in configuration to a driver
+	// for granting access to a bucket
+	// +optional
+	Parameters map[string]string `json:"parameters,omitempty"`
```
Comment (thotz) [inline apis/objectstorage/v1alpha2/types.go:264]: There should be some sort of quota offering in BucketAccessClass as well to restrict users, say if the same bucket is shared among different users which has a limit of 10Gb. If one user consumes 9Gb and otherone will have only 1Gb access remaining

**Review by thotz — CHANGES_REQUESTED** [review]
Any plans to tie up with https://kubernetes.io/docs/concepts/policy/resource-quotas/#storage-resource-quota like PVCs and storageclasses

File: apis/objectstorage/v1alpha2/types.go (hunk 1 above)
Comment (BlaineEXE) [reply apis/objectstorage/v1alpha2/types.go:209]: Is that something that can be controlled via Kubernetes RBAC?

**Comment by xing-yang** [comment]
@BlaineEXE Can you work on a design doc as this is a new feature?

**Comment by xing-yang** [comment]
In Kubernetes, there is a way to specify the number of CRs.  Maybe that is useful as well.  See an example here: https://github.com/kubernetes-csi/external-snapshotter#setting-quota-limits-with-snapshot-custom-resources

**Comment by BlaineEXE** [comment]
Closing this after discussion today. Likely, there is no need for v1alpha2. We can add changes to the v1alpha1 types freely (unlike beta/stable). https://github.com/kubernetes/enhancements/issues/4194

File: apis/objectstorage/v1alpha2/types.go (hunk 2 above)
Comment (BlaineEXE) [reply apis/objectstorage/v1alpha2/types.go:264]: This design is specifically only for bucket quotas. Rook users have given feedback that they want bucket quotas, but they seem to not have strong desires for access/user-level quotas. Given that, and in the interest of making small, incremental changes, I have intentionally omitted BucketAccess quotas from the proposal.

**Comment by BlaineEXE** [comment]
Closing this, as likely a v1alpha2 won't be proposed, instead opting to modify the v1alpha1 api as allowed by "alpha" state.

**Pull request closed without merging**